std/protocols/lookup.asm

```diff
@@ -88,11 +69,14 @@ let compute_next_z: Fp2<expr>, Constr, expr -> fe[] = query |acc, lookup_constra
     
 // Adds constraints that enforce that rhs is the lookup for lhs
 // Arguments:
+// - is_first: A column that is 1 for the first row and 0 for the rest
+// - alpha: A challenge used to compress the LHS and RHS values
```

Comment:
I'm sorry, but this is horrible :(

I can understand the is_first thing, but the challenges? Currently, they have a globally unique ID and it doesn't matter which namespace they are in. You could just as well declare them locally inside the function or use a function that returns a challenge (as en expr).

Reply:
Ok, maybe it's not as horrible as I thought, but structs would greatly improve the readability here.

Reply:
I see your point! Let me expand a bit. For example, with the changes of #1511, the `test_data/std/permutation_via_challenges.asm` test compiles to this optimized PIL:
```
namespace std::convert(8);
    let fe = [];
namespace std::protocols::permutation(8);
    col alpha1 = std::prelude::challenge(0, 1);
    col beta1 = std::prelude::challenge(0, 3);
namespace std::prover(8);
    enum Query {
        Input(int),
        Output(int, int),
        Hint(fe),
        DataIdentifier(int, int),
        None,
    }
namespace main(8);
    col fixed first_four = [1, 1, 1, 1, 0, 0, 0, 0];
    col witness a1(i) query std::prover::Query::Hint(std::convert::fe::<int>(i));
    col witness a2(i) query std::prover::Query::Hint(std::convert::fe::<int>(i + 42));
    col witness b1(i) query std::prover::Query::Hint(std::convert::fe::<int>(7 - i));
    col witness b2(i) query std::prover::Query::Hint(std::convert::fe::<int>(7 - i + 42));
    col witness stage(1) z;
    col fixed is_first(i) { if i == 0 { 1 } else { 0 } };
    main.is_first * (main.z - 1) = 0;
    ((1 - main.first_four) * (std::protocols::permutation::beta1 - (std::protocols::permutation::alpha1 * main.b1 + main.b2) - 1) + 1) * main.z' = (main.first_four * (std::protocols::permutation::beta1 - (std::protocols::permutation::alpha1 * main.a1 + main.a2) - 1) + 1) * main.z;
```
But in #1470, we split by namespace, so the reference to `std::protocols::permutation::alpha1` cannot be resolved. This is the only case where this happen in all our tests.

It seems like challenges are treated like intermediate polynomials. If we split by namespace, it actually fails in [this line](https://github.com/powdr-labs/powdr/blob/91d41df9a06d52be9dc9c7d35756c3606abc67f1/ast/src/analyzed/mod.rs#L383), when inlining intermediate polynomials!

So, I see three long-term solutions:
1. PIL functions can declare challenges in the caller's namespace, just like they already can with phase-0 witness columns (and should with fixed columns). If that's the solution, this PR is not a worse intermediate step than #1511, IMO.
2. The definition of challenges is in-lined earlier, in the condenser run.
3. In #1470, I try to detect this and merge the two namespaces. That's what I did before, but the code is much simpler if we can assume that namespaces are self-contained (except for connecting permutations or lookups).

What do you think? I feel like solution 2 could be good as well, but I can't judge if it is easy to implement.

Reply:
Indeed, `alpha1` and `beta1` are intermediate columns, and that's unrelated to them containing references to challenges. Exactly the same would happen without challenges.

So in order to collect columns for splitting namespaces, intermediate columns should be traversed and only the witness columns should be collected.

I think of intermediate columns as purely symbolic expressions - they don't even have a degree and don't inherit the degree from the namespace they are declared in. If you want a witness column, you would use `let alpha1; alpha1 = ...`.